Remove Shelley style certs in Dijkstra era

Lucsanszky opened the pull request:
# Description
Resolves #4963 
<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [x] Tests added or updated when needed.
- [x] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [ ] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

Lucsanszky committed:
Remove Shelley style certs in Dijkstra era

Lucsanszky committed:
Wire in `ImpTest` skeleton for Dijkstra era

Lucsanszky committed:
Separate tests without certificate deposits

Lucsanszky committed:
Add `registerStakeCredentialWithDeposit`

Lucsanszky committed:
Add `registerPoolWithDeposit`

Lucsanszky committed:
Remove extraneous `ShelleyEraTxCert` constraint

It is not required by the `STS (ShelleyPOOL)` instance

Lucsanszky committed:
Decouple EraTxCerts and remove them from EraImps

* Decouple `ConwayEraTxCert` from `ShelleyEraTxCert`

* Remove `ShelleyEraTxCert` from `ShelleyEraImp` and `ConwayEraTxCert`
from `ConwayEraImp`
  These should not constrain `EraImp`s for the same
  reason they don't constrain `EraTest`s.
  See #4992

* Make some Conway `ImpTest` helpers era generic by only using
certificates with deposits

Lucsanszky committed:
Add `regDelegToDRep` helper

Lucsanszky committed:
Add `registerRewardAccountWithDeposit` helper

Lucsanszky committed:
Adjust Conway Imp tests for Dijkstra era

* Generalise compatible tests by using certificates with deposits
* Move tests that are reliant on certs w/out deposits and only run them pre-Dijkstra

Lucsanszky committed:
Remove `ShelleyEraTxCert` constraint from Imp tests

Lucsanszky committed:
Add `EraSpecificSpec` class

Lucsanszky committed:
Adjust Babbage `UtxowSpec` for Dijkstra era

* Move tests that are reliant on certs w/out deposits and only run them pre-Dijkstra

Lucsanszky committed:
Adjust Alonzo `UtxowSpec` for Dijkstra era

* Move tests that are reliant on certs w/out deposits and only run them pre-Dijkstra

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -302,6 +303,10 @@ instance ShelleyEraImp era => ImpSpec (LedgerSpec era) where
 
+class EraTest era => EraSpecificSpec era where
+  eraSpecific :: SpecWith (ImpInit (LedgerSpec era))
+  eraSpecific = pure ()
```
Let's give it a Spec suffix for clarity and consistency
```suggestion
  eraSpecificSpec :: SpecWith (ImpInit (LedgerSpec era))
  eraSpecificSpec = pure ()
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,344 @@
+data DijkstraDelegCert
+  = DijkstraRegCert !StakeCredential !Coin
+  | DijkstraUnRegCert !StakeCredential !Coin
+  | DijkstraRegDelegCert !StakeCredential !Delegatee !Coin
```
We need to preserve this certificate, otherwise people will not be change their delegation without unregistration
```suggestion
  | DijkstraDelegCert !StakeCredential !Delegatee
  | DijkstraRegDelegCert !StakeCredential !Delegatee !Coin
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,344 @@
-  getDelegTxCert (ConwayTxCertDeleg (ConwayDelegCert cred d)) = Just (cred, d)
-  getDelegTxCert _ = Nothing
+  mkDelegTxCert = notSupportedInThisEra
+  getDelegTxCert = const Nothing
```
We do need to support this certificate.

lehins reviewed (changes requested):
I found one critical mistake that needs to be fixed. You ended up deprecating a certificate that did not need to be deprecated. It is sort of my fault, since ticket did not produce the exact list of certificates that needed deprecation.
In any case, it does need to be fixed. Please let me know when it is resolved and I will do a full PR review.

Lucsanszky committed:
Rename `eraSpecific` method to `eraSpecificSpec`

Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>

Lucsanszky committed:
Reinstate `DelegTxCert` in Dijkstra era

Without this, users won't be able to change their
delegation without unregistration.

Also, re-organise some tests again and run them in each era
if they are general enough (i.e not dependent on Shelley style certs).

Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>

Lucsanszky committed:
Adjust `cardano-ledger-api` `QuerySpec`

Lucsanszky commented:
This is now ready for review again. Once approved, I will fixup the `WIP` commits and perhaps some others too.

Lucsanszky commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs:
```diff
@@ -72,14 +75,16 @@ spec = do
           mkBasicTx mkBasicTxBody
             & bodyTxL . certsTxBodyL
               .~ SSeq.fromList
+                -- NOTE: we might not want to do all this again,
```
Commenting for extra visibility: do we want to run these post-Conway as well (with the necessary adjustments), or should I remove it?

lehins replied:
yes, we do want as many tests as possible for as many eras as possible, so if you could, please do the necessary adjustments.

Lucsanszky replied:
Fortunately, the only necessary adjustment in this case is just to remove the comment then. :) Have this both in the era generic spec (with the required adjustments for post-conway eras) and the era specific spec. Thanks for confirming that we need both!

Lucsanszky commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -87,7 +92,26 @@
 
 ### `testlib`
 
-* Added `Examples` module with: `ledgerExamples`, `exampleConwayCerts`
```
~Took the liberty to modify this entry by removing `exampleConwayCerts`, since only a few commits separates the two, it will probably go into the same release (unless the upcoming one will go ahead without this PR) and because `exampleConwayCerts` were only exported due to the fact that Dijkstra era certificates weren't properly implemented yet.~ Never mind, this PR probably won't be included in the upcoming release.

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs:
```diff
@@ -252,6 +232,46 @@ spec = describe "Invalid transactions" $ do
+  SpecWith (ImpInit (LedgerSpec era))
+alonzoEraSpecificSpec = describe "Invalid transactions" $ do
+  -- let resetAddrWits tx = updateAddrTxWits $ tx & witsTxL . addrTxWitsL .~ []
+  --     fixupResetAddrWits = fixupPPHash >=> resetAddrWits
```
Dead code?
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/DelegSpec.hs:
```diff
@@ -756,8 +713,66 @@ spec = do
+            & bodyTxL . certsTxBodyL .~ [regTxCert]
+        expectRegistered cred
+
+  describe "Delegate stake" $ do
```
I suggest making a copy of this test in shelley that uses shelley only certificates and change this test to not use shelley style certificates and extract it from `conwayEraSpecificSpec`

Lucsanszky replied:
Let me address this in a separate PR. Shelley doesn't have a `DelegSpec`, so I think it would make sense to create a separate PR `Create DelegSpec for Shelley`, where I also resolve this.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -388,6 +394,7 @@ genUnRegTxCert stakingCredential = do
 genRegTxCert ::
```
We can make `genRegTxCert` and `genUnRegTxCert` type class  methods for `ShelleyEraImp` type class.

This would allow us to use these functions and consequently tests that use them in all eras.

Lucsanszky replied:
I have opened a PR #5220 with some similar improvements, so I'd rather do this there.

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+    RegTxCert {} -> Left RegTxCertExpunged
+    -- Using wildcard here instead of a pattern match on UnRegTxCert in order to
+    -- workaround ghc disrespecting the completeness pragma.
+    _ -> Left UnRegTxCertExpunged
```
```suggestion
    UnRegTxCert {} -> Left UnRegTxCertExpunged
    -- Using wildcard here in order to workaround ghc disrespecting
    -- the completeness pragma in presence of `PreviousEra` type family.
    _ -> error "Impossible: all patterns have been accounted for"
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+      ConwayResignCommitteeColdKey coldCred _ -> credKeyHashWitness coldCred
+      ConwayRegDRep cred _ _ -> credKeyHashWitness cred
+      ConwayUnRegDRep cred _ -> credKeyHashWitness cred
+      ConwayUpdateDRep cred _ -> credKeyHashWitness cred
```
You should be able to reuse this from Conway if you extract `govWitness` as a top level function and rename it to `conwayGovCertVKeyWitness`

Lucsanszky replied:
I started factoring this out along with the other `govWitness` function for `ScriptHash`es but obviously those two can't be merged, so I ended up not factoring out anything because I thought this is not worth exporting, despite reducing duplication. Anyways, I don't have a strong opinion on it, so I'm happy to comply.

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+      ConwayUpdateDRep cred _ -> credKeyHashWitness cred
 
-  mkUnRegTxCert c = ConwayTxCertDeleg $ ConwayUnRegCert c SNothing
+instance ShelleyEraTxCert DijkstraEra where
```
This instance for `DijkstraEra` needs to be removed, since all of the certificates that it used to support has been removed

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxInfo.hs:
```diff
@@ -140,6 +149,26 @@ instance EraPlutusTxInfo 'PlutusV1 DijkstraEra where
+  RegDepositTxCert stakeCred _deposit ->
+    Right $ PV1.DCertDelegRegKey (PV1.StakingHash (transCred stakeCred))
+  UnRegDepositTxCert stakeCred _refund ->
+    Right $ PV1.DCertDelegDeRegKey (PV1.StakingHash (transCred stakeCred))
```
We need to provide backwards compatibility for the delegation certificate:
```suggestion
    Right $ PV1.DCertDelegDeRegKey (PV1.StakingHash (transCred stakeCred))
  DelegTxCert stakeCred (DelegStake keyHash) ->
    Just $ PV1.DCertDelegDelegate (PV1.StakingHash (transCred stakeCred)) (transKeyHash keyHash)
```

Previously it was done through `DelegStakeTxCert keyHash`, which is the pattern synonym that is no longer available in Dijkstra, but it is equivalent to `DelegTxCert stakeCred (DelegStake keyHash)`

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxInfo.hs:
```diff
@@ -232,6 +261,39 @@ instance EraPlutusTxInfo 'PlutusV3 DijkstraEra where
 
+transTxCert ::
+  (ConwayEraTxCert era, TxCert era ~ DijkstraTxCert era) => ProtVer -> TxCert era -> PV3.TxCert
+transTxCert pv = \case
```
protocol version is no longer needed in Dijkstra. It was needed in Conway only to fix a bug on an intra-era hardfork: #4571 
```suggestion
transTxCert ::
  (ConwayEraTxCert era, TxCert era ~ DijkstraTxCert era) => TxCert era -> PV3.TxCert
transTxCert = \case
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxInfo.hs:
```diff
@@ -232,6 +261,39 @@ instance EraPlutusTxInfo 'PlutusV3 DijkstraEra where
+    let transDeposit
+          | hardforkConwayBootstrapPhase pv = Nothing
+          | otherwise = Just (transCoinToLovelace deposit)
+     in PV3.TxCertRegStaking (transCred stakeCred) transDeposit
```
See my previous comment
```suggestion
    PV3.TxCertRegStaking (transCred stakeCred) (transCoinToLovelace deposit)
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxInfo.hs:
```diff
@@ -232,6 +261,39 @@ instance EraPlutusTxInfo 'PlutusV3 DijkstraEra where
+    let transRefund
+          | hardforkConwayBootstrapPhase pv = Nothing
+          | otherwise = Just (transCoinToLovelace refund)
+     in PV3.TxCertUnRegStaking (transCred stakeCred) transRefund
```
See my previous comment
Also you lost tralsation of `DelegTxCert` due to the previous mistake of deprecating it
```suggestion
    PV3.TxCertUnRegStaking (transCred stakeCred) (transCoinToLovelace refund)
  DelegTxCert stakeCred delegatee ->
    PV3.TxCertDelegStaking (transCred stakeCred) (transDelegatee delegatee)
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+        <> encCBOR dRep
+    DijkstraRegCert cred deposit ->
+      encodeListLen 3
+        <> encodeWord8 19
```
Using new tags makes this change totally backwards incompatible. We need to continue using the same tags, since certificates are the same.
```suggestion
        <> encodeWord8 7
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+        <> encCBOR deposit
+    DijkstraUnRegCert cred deposit ->
+      encodeListLen 3
+        <> encodeWord8 20
```
```suggestion
        <> encodeWord8 8
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+        <> encCBOR deposit
+    DijkstraRegDelegCert cred (DelegStake poolId) deposit ->
+      encodeListLen 4
+        <> encodeWord8 21
```
```suggestion
        <> encodeWord8 11
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+        <> encCBOR deposit
+    DijkstraRegDelegCert cred (DelegVote drep) deposit ->
+      encodeListLen 4
+        <> encodeWord8 22
```
```suggestion
        <> encodeWord8 12
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+        <> encCBOR deposit
+    DijkstraRegDelegCert cred (DelegStakeVote poolId dRep) deposit ->
+      encodeListLen 5
+        <> encodeWord8 23
```
```suggestion
        <> encodeWord8 13
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+      deposit <- decCBOR
+      pure (n, RegDepositDelegTxCert cred delegatee deposit)
+    {-# INLINE regDelegCertDecoder #-}
+{-# INLINE dijkstraTxCertDelegDecoder #-}
```
It is wrong to change tags for no reason. We must keep backwards compatibility on the serialization level as much as possible!
```suggestion
      | 7 <= t -> conwayTxCertDelegDecoder t
    t -> invalidKey t
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+
+instance
+  ( ConwayEraTxCert era
+  , ShelleyEraTxCert era
```
DijkstraEra will no longer have instance for this class
```suggestion
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+  decCBOR = decodeRecordSum "DijkstraTxCert" $ \case
+    t
+      | 0 <= t && t < 2 -> fail "Certificates without deposits are no longer supported"
+      | t == 2 -> shelleyTxCertDelegDecoder t
```
```suggestion
      | 0 <= t && t < 2 -> fail "Certificates without deposits are no longer supported"
      | t == 2 -> do
        cred <- decCBOR
        stakePool <- decCBOR
        pure (3, DelegTxCert cred (DelegStake stakePool))
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,395 @@
+    {-# INLINE regDelegCertDecoder #-}
+{-# INLINE dijkstraTxCertDelegDecoder #-}
+
+instance (Era era, Val (Value era)) => ToCBOR (DijkstraTxCert era) where
+  toCBOR = toPlainEncoding (eraProtVerLow @era) . encCBOR
```
That is the default implementation
```suggestion
instance (Era era, Val (Value era)) => ToCBOR (DijkstraTxCert era)
```

lehins reviewed (changes requested):
It looks good overall, but there are still some mistakes that need to be resolved.

Lucsanszky committed:
Fix encoding and decoding

Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>

Lucsanszky committed:
Adjust Shelley `PoolSpec` for Dijkstra era

* Move tests that are reliant on certs w/out deposits and only run them pre-Dijkstra

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/DelegSpec.hs:
```diff
@@ -698,45 +672,28 @@ spec = do
         expectNothingExpr (accountState ^. stakePoolDelegationAccountStateL)
 
-    expectDelegatedVote :: HasCallStack => Credential 'Staking -> DRep -> ImpTestM era ()
+    expectDelegatedVote :: Credential 'Staking -> DRep -> ImpTestM era ()
```
Why did you remove those? Could you please reinstate all `HasCallStack` that were removed.
These constraints are really useful, because without them you will not see which occurance of this function is leading to the test failure.
```suggestion
    expectDelegatedVote :: HasCallStack => Credential 'Staking -> DRep -> ImpTestM era ()
```

Lucsanszky replied:
Removed them by accident. :sweat_smile: I was shuffling those around so many times, into and out of a `where` clause, adding and removing the constraints that it seems like at one point I've lost them completely. I realised this in my other PR here: https://github.com/IntersectMBO/cardano-ledger/pull/5220 
Although that pertains to different parts, the `EraImp` methods, where I also missed the `HasCallStack` constraint for the actual era specific functions. Anyways, thanks for catching!

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/Rules/Cert.hs:
```diff
@@ -29,3 +45,71 @@ instance InjectRuleFailure "CERT" ShelleyPoolPredFailure DijkstraEra where
   injectFailure = GovCertFailure
+
+instance
+  forall era.
```
Redundant
```suggestion
```

lehins commented on eras/dijkstra/src/Cardano/Ledger/Dijkstra/TxCert.hs:
```diff
@@ -1,135 +1,346 @@
+      encodeListLen 3
+        <> encodeWord8 8
+        <> encCBOR cred
+        <> encCBOR deposit
```
Not necessary, but I think this way it will be a bit nicer.
Swapping the order makes the tags increase, instead of being out of order.
```suggestion
    DijkstraRegCert cred deposit ->
      encodeListLen 3
        <> encodeWord8 7
        <> encCBOR cred
        <> encCBOR deposit
    DijkstraUnRegCert cred deposit ->
      encodeListLen 3
        <> encodeWord8 8
        <> encCBOR cred
        <> encCBOR deposit
    DijkstraDelegCert cred (DelegStake poolId) -> encodeShelleyDelegCert $ ShelleyDelegCert cred poolId
    DijkstraDelegCert cred (DelegVote drep) ->
      encodeListLen 3
        <> encodeWord8 9
        <> encCBOR cred
        <> encCBOR drep
    DijkstraDelegCert cred (DelegStakeVote poolId dRep) ->
      encodeListLen 4
        <> encodeWord8 10
        <> encCBOR cred
        <> encCBOR poolId
        <> encCBOR dRep
```

Lucsanszky replied:
Agreed, fair point!

lehins reviewed (approved):
Beautiful!
Thank you!

Lucsanszky committed:
Apply suggestions from code review

Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>

teodanciu commented:
@Lucsanszky  Could you explain, for the benighted - what the deal is with `EraSpecificSpec`  class ? 
I can see that each era accumulates the "era specific" ones from the previous ones, 
for alonzo: `shelleyEraSpecificSpec >> alonzoEraSpecificSpec`, 
for babbage: `shelleyEraSpecificSpec >> alonzoEraSpecificSpec >>  babbageEraSpecificSpec`,
for conway:  `shelleyEraSpecificSpe >> alonzoEraSpecificSpec >> babbageEraSpecificSpec >> conwayEraSpecificSpec`

Is the semantic of `eraSpecificSpec`:  the tests in these functions should be used in future eras? So are they technically specific for all eras?  How does this relate to how we are already including the `Imp` from  previous eras when running Imp tests? 
(for example, Babbage `Imp` module calls `AlonzoImp.spec @era` when defining spec)  

And where is the `eraSpecific` method used?  I can see it used in shelley `Imp`, but not anywhere else,  but perhaps I'm just missing it.

Lucsanszky commented:
> @Lucsanszky Could you explain, for the benighted - what the deal is with `EraSpecificSpec` class ? I can see that each era accumulates the "era specific" ones from the previous ones, for alonzo: `shelleyEraSpecificSpec >> alonzoEraSpecificSpec`, for babbage: `shelleyEraSpecificSpec >> alonzoEraSpecificSpec >> babbageEraSpecificSpec`, for conway: `shelleyEraSpecificSpe >> alonzoEraSpecificSpec >> babbageEraSpecificSpec >> conwayEraSpecificSpec`

The raison d'etre of the `EraSpecificSpec` class is to allow us to conveniently select `spec`s and tests to run per era. It is @lehins's brainchild and it was necessitated by deprecating certificates without deposits post-Conway. Pre-Conway, certificates didn't have deposits, Conway introduced certificates with deposits and post-Conway we will only allow certificates with deposits. This caused some calamity in our current `Imp` test setup, where in each era we also run the `Imp` tests from previous eras. My first iteration merely ripped out the parts that breaks post-Conway and made sure that we don't run certain tests in certain eras. However, this is not a well-organised solution and is prone to oversight in the long-run. This typeclass approach suggested by @lehins is clearer and it is also a well established pattern in our codebase. 
> 
> Is the semantic of `eraSpecificSpec`: the tests in these functions should be used in future eras? So are they technically specific for all eras? How does this relate to how we are already including the `Imp` from previous eras when running Imp tests? (for example, Babbage `Imp` module calls `AlonzoImp.spec @era` when defining spec)

I must admit, `EraSpecificSpec` was a very poor choice of name on my part. Fortunately, @lehins is already cooking a PR which among other things will also address the unfortunate naming. To answer your question, the semantic is more along the lines of what I described above: you can select which tests (collected by the functions) to run for the given era, when providing an instance for that era. Essentially, we will be able to mix and match conveniently in an organised way from here on out. I think the best way to explain or demonstrate what this does is by comparing the `EraSpecificSpec` instance of `ConwayEra` and `DijkstraEra`. In case of the `ConwayEra` we run all the `eraSpecificSpec`s  from previous eras, whereas in `DijkstraEra` we stop running those tests altogether, essentially saying: "Only run these tests up to Conway". However, it is worth noting that with the current setup, we could also say things like "Only run tests from Shelley and Alonzo". Surely enough, we could achieve this behaviour without creating a typeclass for it, but this typeclass solution yields better organisation. Not to mention, that the compiler will yell at us if we forget to provide an instance for an era. 
> 
> And where is the `eraSpecific` method used? I can see it used in shelley `Imp`, but not anywhere else, but perhaps I'm just missing it.

Yes, because we only need to use it there. As you pointed out, each Imp spec calls the previous era's Imp spec. This, combined with the fact that we are providing `EraSpecificSpec` instances for all the eras yields that even when we are in `Conway` for example, `eraSpecificSpec` will be evaiuated but it will pick the implementation defined in the `EraSpecificSpec ConwayEra` instance, which in our case is: 
```
  eraSpecificSpec =
    ShelleyImp.shelleyEraSpecificSpec
      >> AlonzoImp.alonzoEraSpecificSpec
      >> BabbageImp.babbageEraSpecificSpec
      >> conwayEraSpecificSpec
```
